client/client.go

```diff
@@ -1,29 +1,401 @@
+	MaxPending int
+	// The timeout to apply when establishing new connections.
+	ConnectTimeout time.Duration
+	// The timeout to apply when waiting for incoming responses.
+	ReadTimeout time.Duration
 }
 
-func NewCqlClient(address string, codec frame.RawCodec) *CqlClient {
+// Creates a new CqlClient with default options. Leave credentials nil to opt out from authentication.
+func NewCqlClient(remoteAddress string, credentials *AuthCredentials) *CqlClient {
 	return &CqlClient{
-		address: address,
-		codec:   codec,
+		RemoteAddress:  remoteAddress,
+		Credentials:    credentials,
+		MaxInFlight:    DefaultMaxInFlight,
+		MaxPending:     DefaultMaxPending,
+		ConnectTimeout: DefaultConnectTimeout,
+		ReadTimeout:    DefaultReadTimeout,
 	}
 }
 
-func (client *CqlClient) Connect() (*CqlConnection, error) {
-	c, err := net.Dial("tcp", client.address)
-	if err != nil {
+func (client *CqlClient) String() string {
+	return fmt.Sprintf("CQL client [%v]", client.RemoteAddress)
+}
+
+// Connect establishes a new TCP connection to the client's remote address. The returned CqlClientConnection is ready
+// to use, but one must initialize it manually, for example by calling CqlClientConnection.InitiateHandshake.
+// Alternatively, use ConnectAndInit to get a fully-initialized connection.
+func (client *CqlClient) Connect() (*CqlClientConnection, error) {
+	log.Debug().Msgf("%v: connecting", client)
+	if conn, err := net.DialTimeout("tcp", client.RemoteAddress, client.ConnectTimeout); err != nil {
+		return nil, fmt.Errorf("%v: cannot establish TCP connection: %w", client, err)
+	} else {
+		connection, err := newCqlClientConnection(
+			conn,
+			client.Credentials,
+			client.Codec,
+			client.MaxInFlight,
+			client.MaxPending,
+			client.ReadTimeout,
+		)
+		log.Info().Msgf("%v: new TCP connection established: %v", client, connection)
+		return connection, err
+	}
+}
+
+// ConnectAndInit establishes a new TCP connection to the server, then initiates a handshake procedure using the
+// specified protocol version. The CqlClientConnection connection will be fully initialized when this method returns.
+func (client *CqlClient) ConnectAndInit(version primitive.ProtocolVersion) (*CqlClientConnection, error) {
+	if connection, err := client.Connect(); err != nil {
 		return nil, err
+	} else {
+		return connection, connection.InitiateHandshake(version)
+	}
+}
+
+// CqlClientConnection encapsulates a TCP client connection to a remote Cassandra-compatible backend.
+// CqlClientConnection instances should be created by calling CqlClient.Connect or CqlClient.ConnectAndInit.
+type CqlClientConnection struct {
+	conn            net.Conn
+	codec           frame.Codec
+	readTimeout     time.Duration
+	credentials     *AuthCredentials
+	inFlightHandler *inFlightRequestsHandler
+	outgoing        chan *frame.Frame
+	events          chan *frame.Frame
+	waitGroup       *sync.WaitGroup
+	closed          int32
+}
+
+func newCqlClientConnection(
```

Comment:
Can we expose this function? We have a use case where we use an already open `net.Conn` as a CQL connection and exposing this "constructor" would enable that use case